controller/rollout_load_controller.go

```diff
@@ -0,0 +1,1579 @@
+		log = log.WithField("backupTargetURL", c.backupTargetURL)
+	}
+
+	if c.engineImage != "" {
+		log = log.WithField("engineImage", c.engineImage)
+	}
+
+	return log
+}
+
+func (c *RolloutLoadController) handleRolloutErr(rollout *longhorn.Rollout, reason, message string, err error, log logrus.FieldLogger) {
+	rollout.Status.State = longhorn.RolloutStateError
+	rollout.Status.Conditions = types.SetCondition(
+		rollout.Status.Conditions,
+		longhorn.RolloutConditionTypeError,
+		longhorn.ConditionStatusTrue,
+		reason,
+		fmt.Sprintf("%v: %v", message, err),
+	)
+
+	log.WithError(err).Error(message)
+
+	c.eventRecorder.Eventf(rollout, corev1.EventTypeWarning, EventReasonFailed, util.ConvertToSentence(message))
+}
+
+func (c *RolloutLoadController) rolloutLoad() (err error) {
+	log := c.getLoggerForRolloutLoad()
+	errConditionReason := ""
+	errConditionMessage := ""
+	nextState := longhorn.RolloutStateUnknown
+	defer func() {
+		rolloutErr := err
+
+		rollout, err := c.ds.GetRollout(c.rollout.Name)
+		if err != nil {
+			log.WithError(err).Error("failed to get rollout")
+			close(c.stopCh)
+		}
+
+		if errConditionReason+errConditionMessage != "" {
+			c.handleRolloutErr(rollout, errConditionReason, errConditionMessage, rolloutErr, log)
+		} else {
+			rollout.Status.State = nextState
+		}
+
+		rollout, err = c.ds.UpdateRolloutStatus(rollout)
+		if err != nil {
+			log.Warnf(RolloutLoadMgsEnqueueFmt, RolloutLoadErrorUpdateStatusFailed)
+			c.enqueue()
+			return
+		}
+
+		if rollout.Status.State != longhorn.RolloutStateCompleted && rollout.Status.State != longhorn.RolloutStateError {
+			log.Infof(RolloutLoadMgsEnqueueFmt, rollout.Status.State)
+			c.enqueue()
+			return
+		}
+
+		_, err = c.ds.RemoveRolloutLabel(rollout)
+		if err != nil {
+			log.WithError(err).Warnf(RolloutLoadMgsEnqueueFmt, "failed to remove rollout label")
+			c.enqueue()
+			return
+		}
+
+		log.Info("Shutting down")
+		close(c.stopCh)
+	}()
+
+	rollout, err := c.ds.GetRollout(c.rollout.Name)
+	switch rollout.Status.State {
+	case longhorn.RolloutStatePending:
+		_, err = c.initializeRollout(log)
+		if err != nil {
+			errConditionReason = longhorn.RolloutConditionReasonInitialize
+			errConditionMessage = "failed to initialize rollout"
+		}
+
+		nextState = longhorn.RolloutStateDownloading
+		return
+	case longhorn.RolloutStateDownloading:
+		_, err = c.Download(log)
+		if err != nil {
+			errConditionReason = longhorn.RolloutConditionReasonDownload
+			errConditionMessage = "failed to download rollout from backup target"
+		}
+
+		nextState = longhorn.RolloutStateUnpacking
+		return
+	case longhorn.RolloutStateUnpacking:
+		_, err = c.Unpack(log)
+		if err != nil {
+			errConditionReason = longhorn.RolloutConditionReasonUnpack
+			errConditionMessage = "failed to unpack rollout from file"
+		}
+
+		nextState = longhorn.RolloutStateRestoring
+		return
+	case longhorn.RolloutStateRestoring:
+		wg := &sync.WaitGroup{}
+		restoreFns := map[string]func() error{
```

Comment:
I am thinking if we can restore items by sequence.
For example, if we restore volume then settings, applying dangerous setting would crash the restoring volumes. Or restoring PV & PVC before volumes looks weird as well.

Reply:
IIUC restoring volumes will be attached, and the webhook validator prohibits updating some settings while volumes are attached, no?
If this is true, then when ` c.restore("SettingList", c.restoreSettings, log, wg)`:
1. `c.restoreSettings` returns error
1. c.restore retrys until either `c.restoreSettings()` success or timeout.

If this is not true, then we can restore the settings before other resources.

I think restoring items by sequence will make restoration hard to maintain, as we have to be mindful of the sequence in the future. Instead, the idea is to let each resource restore fn() be responsible for its own dependency handling. 

For example:
- `restorePersistentVolumes()` will check its depending Volume and returns error if it doesn't exist yet.
- `restorePersistentVolumeClaims()` checks its depending PersistentVolume and returns error if it doesn't exist yet.
- The caller c.restore() will be responsible for the retry and timeout.

Reply:
> IIUC restoring volumes will be attached, and the webhook validator prohibits updating some settings while volumes are attached, no? What am I missing?

Probably NO... Will the validator block the instance manager CPU settings update when there are attached volumes?

> I think restoring items by sequence will make restoration hard to maintain, as we have to be mindful of the sequence in the future. Instead, the idea is to let each resource restore fn() be responsible for its own dependency handling.

Definitely agree.

Reply:
> Probably NO... Will the validator block the instance manager CPU settings update when there are attached volumes?

Ok, I guess we didn't block it because we weren't expecting the user to restore volume and change CPU settings at the same time...?

Moved the restore setting out and handle it first.

Reply:
> I guess we didn't block it because we weren't expecting the user to restore volume and change CPU settings at the same time...?

Here, YES the concern is that the CPU change would lead to IM pod restart then interrupting the restoring. But I am not sure the reason why Longhorn does not block the setting update when there are running volumes/processes.
 
Actually, we already considered delaying the CPU changes then applying them to the IM pods only when there is no process running on the pod.